### DISCUSSION TITLE
JSON request remove parent for key sql data model

### DISCUSSION THREAD

**Discussion by lluked** [fastapi/fastapi#12511] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class Ticket(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True) 
    subject: str
    comment: str
    tags: str

@app.post("/tickets/")
async def create_ticket(ticket: Ticket, session: SessionDep) -> Ticket:
    db_ticket = Ticket.model_validate(ticket)
    session.add(db_ticket)
    session.commit()
    session.refresh(db_ticket)
    return db_ticket
```

### Description

I'm looking to access all elements within ticket in the following json object

```
{
    "ticket": {
        "subject": "Test-Ticket",
        "comment": {
            "html_body": "<h2>Test Ticket</h2>"
        },
        "tags": [
            "test-ticket"
        ]
    }
}
```

as if the ticket did not exist and the data was 

```
{
        "subject": "Test-Ticket",
        "comment": {
            "html_body": "<h2>Test Ticket</h2>"
        },
        "tags": [
            "test-ticket"
        ]
 }
```

then use the SQLModel to store in an sql database file, I cannot find any documentation mentioning this. The only examples I have found show when there is no parent key or are using BaseModel from pydantic to have a parent class  but I'm unsure how this would work with sql. This may be a stupid question but I cannot figure it out.

### Operating System

macOS

### FastAPI Version

0.115.2

### Pydantic Version

2.9.2

### Python Version

Python 3.12.7

**Comment by YuriiMotov** [comment]
Is this what you need?

```
from fastapi import FastAPI
from sqlmodel import SQLModel, Field

class Ticket(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True, exclude=True) 
    subject: str
    comment: str
    tags: str

class TicketWrapped(SQLModel):
    ticket: Ticket

app = FastAPI()

@app.post("/tickets/")
async def create_ticket(ticket: Ticket) -> TicketWrapped:
    db_ticket = Ticket.model_validate(ticket)
    return {"ticket": db_ticket}

```

**Reply by lluked** [reply]
Thank you, I thought i tried that and had errors, that must have been from something else

**Reply by lluked** [reply]
> Is this what you need?
> 
> ```
> from fastapi import FastAPI
> from sqlmodel import SQLModel, Field
> 
> class Ticket(SQLModel, table=True):
>     id: int | None = Field(default=None, primary_key=True, exclude=True) 
>     subject: str
>     comment: str
>     tags: str
> 
> class TicketWrapped(SQLModel):
>     ticket: Ticket
> 
> app = FastAPI()
> 
> @app.post("/tickets/")
> async def create_ticket(ticket: Ticket) -> TicketWrapped:
>     db_ticket = Ticket.model_validate(ticket)
>     return {"ticket": db_ticket}
> ```

Had another go and still can't get it to work, using heroes as an example like the docs do:

main.py
```
from typing import Annotated

from fastapi import Depends, FastAPI, Query
from sqlmodel import Field, Session, SQLModel, create_engine, select


class Hero(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    name: str = Field(index=True)

class HeroWrapped(SQLModel):
    hero: Hero


sqlite_file_name = "hero.db"
sqlite_url = f"sqlite:///{sqlite_file_name}"

connect_args = {"check_same_thread": False}
engine = create_engine(sqlite_url, connect_args=connect_args)


def create_db_and_tables():
    SQLModel.metadata.create_all(engine)

def get_session():
    with Session(engine) as session:
        yield session


SessionDep = Annotated[Session, Depends(get_session)]


app = FastAPI()


@app.on_event("startup")
def on_startup():
    create_db_and_tables()


# {'name': 'batman'}
@app.post("/heroes")
def create_hero(hero: Hero, session: SessionDep):
    db_hero = Hero.model_validate(hero)
    session.add(hero)
    session.commit()
    session.refresh(hero)
    return db_hero

# {'hero': {'name': 'robin'}}
@app.post("/heroes_")
def create_hero(hero: Hero, session: SessionDep) -> HeroWrapped:
    db_hero = hero
    # db_hero = Hero.model_validate(hero)
    session.add(hero)
    session.commit()
    session.refresh(hero)
    return db_hero

# Get heroes
@app.get("/heroes")
def read_heroes(
    session: SessionDep,
    offset: int = 0,
    limit: Annotated[int, Query(le=100)] = 100,
) -> list[Hero]:
    heroes = session.exec(select(Hero).offset(offset).limit(limit)).all()
    return heroes
```

create_heroes.py
```
import requests
import json

url_hero = 'http://127.0.0.1:8000/heroes'
url_wrapped_hero = url_hero + '_'
headers = {'Content-Type': 'application/json'}

# {'name': 'batman'}
hero = {'name': 'batman'}
hero_json_data = json.dumps(hero)
# print(hero_json_data)
create_hero_response = requests.post(url_hero, data=hero_json_data, headers=headers)
# print(create_hero_response.json())

# {'hero': {'name': 'batman'}}
hero_wrapped_json_data = json.dumps({'hero': {'name': 'robin'}})
# print(hero_wrapped_json_data)
create_hero_wrapped_response = requests.post(url_wrapped_hero, data=hero_wrapped_json_data, headers=headers)
# print(create_hero_wrapped_response.json())

# Get heroes
get_heroes_response = requests.get(url_hero, headers=headers)
print(get_heroes_response.json())

```

Error
```
sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) NOT NULL constraint failed: hero.name
[SQL: INSERT INTO hero (name) VALUES (?)]
[parameters: (None,)]
(Background on this error at: https://sqlalche.me/e/20/gkpj)
```

**Reply by YuriiMotov** [reply, marked as answer]
To make your `create_heroes.py` script working you need to define input parameter of `/heroes_` endpoint as `HeroWrapped` and then take hero's data from `hero.hero`.

Also, since your return model is `HeroWrapped` you need to return not just `db_hero` but `{"hero": db_hero}`.

Here is a working code:
```
@app.post("/heroes_")
def create_hero(hero: HeroWrapped, session: SessionDep) -> HeroWrapped:
    db_hero = hero.hero
    session.add(db_hero)
    session.commit()
    session.refresh(db_hero)
    return {"hero": db_hero}
```

**Reply by lluked** [reply]
Thanks that sorted it
